viostor/virtio_stor_hw_helper.c

```diff
@@ -194,30 +195,29 @@ RhelDoReadWrite(PVOID DeviceExtension,
     vq = adaptExt->vq[QueueNumber];
     RhelDbgPrint(TRACE_LEVEL_VERBOSE, " QueueNumber 0x%x vq = %p\n", QueueNumber, vq);
 
-    element = &adaptExt->processing_srbs[QueueNumber];
     VioStorVQLock(DeviceExtension, MessageId, &LockHandle, FALSE);
-    if (virtqueue_add_buf(vq,
-                     &srbExt->sg[0],
-                     srbExt->out, srbExt->in,
-                     &srbExt->vbr, va, pa) >= 0) {
+    res = virtqueue_add_buf(vq,
+        &srbExt->sg[0],
+        srbExt->out, srbExt->in,
+        &srbExt->vbr, va, pa);
+
+    if (res >= 0) {
         notify = virtqueue_kick_prepare(vq);
+        element = &adaptExt->processing_srbs[QueueNumber];
         InsertTailList(&element->srb_list, &srbExt->vbr.list_entry);
         element->srb_cnt++;
-        VioStorVQUnlock(DeviceExtension, MessageId, &LockHandle, FALSE);
 #ifdef DBG
         InterlockedIncrement((LONG volatile*)&adaptExt->inqueue_cnt);
 #endif
         result = TRUE;
-    }
```

Comment:
Moving the `element = &adapt->processing_srbs[QueueNumber];` to after the `virtqueue_kick_prepare()` might be an issue here... I will refactor and retest...

Reply:
This is the same for [vioscsi]... I will check there first.

Reply:
Already done in my present [vioscsi] WIP... Will look at [viostor] once [vioscsi] is stable.

```c
    //
    // Moved ouside conditional so (res != VQ_ADD_BUFFER_SUCCESS) can access it for RemoveTailList(&element->srb_list)
    element = &adaptExt->processing_srbs[index];
    //
    // res, i.e. virtqueue_add_buf() will only
    // return 0 () or 28 (ENOSPC), i.e. No space left on device.
    // Therefore setting comparator from >= to ==
    // Also defined VQ_ADD_BUFFER_SUCCESS (as 0) in helper.h
    if (res == VQ_ADD_BUFFER_SUCCESS) {
        //
        // Kicker returns like this:
        // 
        //  if (_vq->vdev->event_suppression_enabled) {
        //      return wrap_around || (bool)vring_need_event(vring_avail_event(&vq->vring), new, old);
        //  } else {
        //      return !(vq->vring.used->flags & VIRTQ_USED_F_NO_NOTIFY);
        //  }
        //
        /* The Host uses this in used->flags to advise the Guest: don't kick me when
         * you add a buffer.  It's unreliable, so it's simply an optimization.  Guest
         * will still kick if it's out of buffers.
         * 
         #define VIRTQ_USED_F_NO_NOTIFY	1
         */
        // So ....
        //
        notify = virtqueue_kick_prepare(adaptExt->vq[QueueNumber]);
        RhelDbgPrint(TRACE_LEVEL_RESERVED8, " After KickPrep Notify State  : %d \n", notify);
        //
        // Moved ouside conditional so else can access
        //element = &adaptExt->processing_srbs[index];
        InsertTailList(&element->srb_list, &srbExt->list_entry);
        element->srb_cnt++;
    } else {
        //
        // So to reiterate:
        // This only gets hit if virtqueue_add_buf() returns 28, i.e. the buffer is full.
        //
        // Put the DbgPrint above rest to ensure we get problem data. Also split the long line.
        RhelDbgPrint(TRACE_LEVEL_FATAL, " Could not put an SRB into a VQ, so complete it with SRB_STATUS_BUSY. QueueNumber = %d, SRB = 0x%p, Lun = %d, TimeOut = %d.\n", \
            QueueNumber, srbExt->Srb, SRB_LUN(Srb), Srb->TimeOutValue);
        //
        // Adding equivalent for InsertTailList() as the 
        // working version 100.85.104.20800 driver
        // had ExInterlockedInsertHeadList() here
        RemoveTailList(&element->srb_list);
        ScsiStatus = SCSISTAT_QUEUE_FULL;
        SRB_SET_SRB_STATUS(Srb, SRB_STATUS_BUSY);
        SRB_SET_SCSI_STATUS(Srb, ScsiStatus);
        StorPortBusy(DeviceExtension, 10);
        CompleteRequest(DeviceExtension, Srb);
        //
        // Don't think these are necessary as the buffer is full.
        //notify = TRUE;
        //virtqueue_notify(adaptExt->vq[QueueNumber]);
        //
        // ...but do we need this...? CompleteRequest() takes 
        // the SRB and finalises it, so best to wipe ours methinks...
        // 
        SrbExt = NULL;
    }
```